src/QueryBuilder.php

```diff
@@ -2,94 +2,143 @@
+    protected $subject;
+
+    public function __construct($subject, ?Request $request = null)
+    {
+        $this->initializeSubject($subject)
+            ->initializeRequest($request ?? app(Request::class));
+    }
+
+    protected function initializeSubject($subject): self
     {
-        parent::__construct(clone $builder->getQuery());
+        throw_unless(
+            $subject instanceof EloquentBuilder || $subject instanceof Relation,
+            InvalidSubject::make($subject)
+        );
+
+        $this->subject = $subject;
 
-        $this->initializeFromBuilder($builder);
+        return $this;
+    }
 
+    protected function initializeRequest(?Request $request = null): self
+    {
         $this->request = $request
             ? QueryBuilderRequest::fromRequest($request)
             : app(QueryBuilderRequest::class);
+
+        return $this;
     }
 
-    /**
-     * Create a new QueryBuilder for a request and model.
-     *
-     * @param string|\Illuminate\Database\Eloquent\Builder|\Illuminate\Database\Eloquent\Relations\Relation $baseQuery Model class or base query builder
-     * @param \Illuminate\Http\Request $request
-     *
-     * @return \Spatie\QueryBuilder\QueryBuilder
-     */
-    public static function for($baseQuery, ?Request $request = null): self
+    public function getEloquentBuilder(): EloquentBuilder
     {
-        if (is_string($baseQuery)) {
-            /** @var Builder $baseQuery */
-            $baseQuery = $baseQuery::query();
+        if ($this->subject instanceof EloquentBuilder) {
+            return $this->subject;
         }
 
-        return new static($baseQuery, $request ?? app(Request::class));
+        if ($this->subject instanceof Relation) {
+            return $this->subject->getQuery();
+        }
+
+        throw InvalidSubject::make($this->subject);
+    }
+
+    public function getSubject()
+    {
+        return $this->subject;
     }
 
     /**
-     * {@inheritdoc}
+     * @param EloquentBuilder|Relation|string $subject
+     * @param Request|null $request
+     * @return static
      */
-    public function get($columns = ['*'])
+    public static function for($subject, ?Request $request = null): self
     {
-        $results = parent::get($columns);
+        if (is_subclass_of($subject, Model::class)) {
+            $subject = $subject::query();
+        }
+
+        return new static($subject, $request);
+    }
 
-        if ($this->request->appends()->isNotEmpty()) {
-            $results = $this->addAppendsToResults($results);
+    public function __call($name, $arguments)
+    {
+        $result = $this->forwardCallTo($this->subject, $name, $arguments);
+
+        if ($result === $this->subject) {
+            return $this;
+        }
+
+        if ($result instanceof Model) {
+            $this->addAppendsToResults(collect([$result]));
+        }
+
+        if ($result instanceof Collection) {
+            $this->addAppendsToResults($result);
         }
```

Comment:
Assuming all of this is gonna be in a v3 of the query builder, I wouldn't mind dropping support for appends. I don't think it's used a lot and I don't like the way it's implemented. Users can always append manually when they get a set of results.

Reply:
@hivokas A little edge case for the forwarded $results as a Paginator instead of collection, I've made #463.
@AlexVanderbist Curious what alternative to appends might be? I'm using them quite alot when the attribute is 'expensive' to always append or makes things too dirty/complex.